File: palindromize.cpp
/****************************************
    
    KMP algorithm

    Input : string

    Output : Min Palindrome length

    Makes Palindrome by using KMP algorithm

    8/23 coded by bmkim
    
****************************************/
#include <vector>
#include <cstring>
#include <iostream>
#include <algorithm>
using namespace std;
vector<int> getPartialMatch(string& s){
    int k = s.size();
    vector<int> ret(k, 0);
    int begin = 1, length = 0;
    while(begin < k){
        if(s[begin] == s[length]){
            length++;
            ret[begin] = length;
            begin++;
        }
        else{
            if(length == 0){
                begin++;
            }
            else{
                length = ret[length-1];
            }
        }
    }
    return ret;
}
int palindromize(string &rev){
    string s = rev;
    reverse(rev.begin(), rev.end());
    vector<int> gpm = getPartialMatch(rev);
    int begin = 0, matched = 0;
    int n = s.size();
    while(begin < n){
        if(matched < n && s[begin+matched] == rev[matched]){
            matched++;
            if(begin + matched == n)
                return matched + begin * 2;
        }
        else{
            if(matched == 0)
                begin++;
            else{
                begin += matched - gpm[matched-1];
                matched = gpm[matched-1];
            }
        }
    }
    return 0;
}
int main(){
    int C;
    cin >> C;
    while(C--){
        string s;
        cin >> s;
        cout << palindromize(s) << endl;
    }
}

File: 종만북/palindromize.py
"""
    
    KMP algorithm

    Input : string

    Output : Min Palindrome length

    Makes Palindrome by using KMP algorithm

    8/23 coded by bmkim
    
"""

def getPartialMatch(word):
    pi = [0] * len(word)
    
    leng = 0
    begin = 1
    while begin < len(word):
        if word[begin] == word[leng]:
            leng += 1
            pi[begin] = leng
            begin += 1
        else:
            if leng != 0:
                leng = pi[leng-1]
            else:
                begin += 1
    return pi

def palindromize(s):
    rev = s[::-1]
    gpm = getPartialMatch(rev)
    begin = 0
    matched = 0
    while begin < len(s):
        if matched < len(s) and s[begin+matched] == rev[matched]:
            matched += 1
            if begin + matched == len(s):
                return matched + begin * 2
        else:
            if matched == 0:
                begin += 1
            else:
                begin += matched - gpm[matched-1]
                matched = gpm[matched-1]
    return 0

if __name__ == "__main__":
    C = int(input())
    for i in range(C):
        s = input()
        print(palindromize(s))
